corehq/apps/reports/views.py

```diff
@@ -798,21 +824,33 @@ def post(self, request, *args, **kwargs):
 def delete_scheduled_report(request, domain, scheduled_report_id):
     user = request.couch_user
     try:
-        rep = ReportNotification.get(scheduled_report_id)
+        scheduled_report = ReportNotification.get(scheduled_report_id)
     except ResourceNotFound:
         # was probably already deleted by a fast-clicker.
         pass
     else:
-        if user._id != rep.owner._id and not user.is_domain_admin(domain):
-            return HttpResponseBadRequest()
+        if not _can_delete_scheduled_report(scheduled_report, user, domain):
+            raise Http404()
 
-        rep.delete()
+        scheduled_report.delete()
         messages.success(request, "Scheduled report deleted!")
     return HttpResponseRedirect(reverse("reports_home", args=(domain,)))
 
 
+def _can_delete_scheduled_report(report, user, domain):
+    if report.domain != domain:
+        return False
+
+    if user._id != report.owner_id and not user.is_domain_admin(domain):
+        return False
+
+    return True
```

Comment:
I think framing these conditions in the positive makes the method easier to understand.
```suggestion
    return (
        report.domain == domain
        and user._id == report.owner_id
        and user.is_domain_admin(domain)
    )
```

Reply:
I think this one is a bit more subjective. The suggestion doesn't recreate the same logic, as you need to either be an owner or a domain admin, not both. I could rewrite it to be `return domain == domain and (owner or domain_admin)`, but I don't know that that is actually clearer. When I was viewing the condition in the negative, it was clear that failing on a domain check (horizontal escalation) was different than failing due to user permissions (vertical escalation), hence the separation.

Reply:
You are correct, I made an error in the logic conversion (should have written it out). The suggestion should have been
```py
    return (
        report.domain == domain
        and (user._id == report.owner_id or user.is_domain_admin(domain))
    )
```